libpromises/rlist.c

```diff
@@ -1206,35 +1204,35 @@ void RlistWrite(Writer *writer, const Rlist *list)
     WriterWriteChar(writer, '}');
 }
 
-/* Note: only single quotes are escaped, as they are used in RlistWrite to
-   delimit strings. If double quotes would be escaped, they would be mangled by
-   RlistParseShown */
-
-static void ScalarWrite(Writer *w, const char *s)
+void ScalarWrite(Writer *writer, const char *s, bool dontQuote)
 {
+    if(!dontQuote)
+        WriterWriteChar(writer, '"');
     for (; *s; s++)
     {
-        if (*s == '\'')
+        if (*s == '"')
         {
-            WriterWriteChar(w, '\\');
+            WriterWriteChar(writer, '\\');
         }
-        WriterWriteChar(w, *s);
+        WriterWriteChar(writer, *s);
     }
+    if(!dontQuote)
+        WriterWriteChar(writer, '"');
 }
 
-void RvalWrite(Writer *writer, Rval rval)
+void RvalWrite(Writer *writer, Rval rval, bool dontQuote)
```

Comment:
In most places we don't quote, and the extra bool is an annoying and somewhat confusing argument to include in every call. Therefore I suggest that you split this into `RvalWrite` and `RvalWriteQuoted`, both taking the same arguments as before the patch. The patch will be much smaller, since you won't have to change every place that uses it.